[dtensor] refactor view ops to use OpStrategy

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126359
* __->__ #126011

As titled. Some ops require adjustment of output shape argument. In rule-based sharding prop, global output shape was inferred in the rule (in `view_ops.py`). In strategy-based sharding prop, it is now obtained from propagated out_tensor_meta (in `sharding_prop.py`).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang @d4l3k

tianyu-l committed:
[dtensor] refactor view ops to use OpStrategy

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] refactor view ops to use OpStrategy"


Previously, the rule-based view op sharding prop adjust a non-tensor arg `local_out_shape` within the rule itself. This was not viable in strategy-based sharding prop.

Thus, this PR is adding a new option `non_tensor_arg_suggestions` into `PlacementStrategy` to address this problem. It also benefits the new factory ops in that we no longer need to compute their local shape and stride in `sharding_prop.py` in a customized way. Instead, we compute all such expected **tensor and non-tensor args** in `tensor_ops.py` and `view_ops.py`, which keeps the `sharding_prop.py` clean.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/ops/view_ops.py:
```diff
@@ -631,121 +609,111 @@ def get_dim_size(cmd: DimSpec) -> Tuple[int, Optional[InputDim]]:
+        for input_placement_strategy in input_strategy.strategies:
+            input_src_spec = input_placement_strategy.output_spec
+
+            with disable_proxy_modes_tracing(), unset_fake_temporarily():
```
why we still need to `propagate_shape_and_sharding`?

I think we want to probably rewrite this function a bit to make sure:
1. we don't create tensors inside this function, maybe replace it with double list
2. no need to compute global output shape, as it's been computed by fake tensor prop outside anyways

This way we should be able to get rid of those context managers

wanchaol commented on torch/distributed/_tensor/ops/view_ops.py:
```diff
@@ -631,121 +609,111 @@ def get_dim_size(cmd: DimSpec) -> Tuple[int, Optional[InputDim]]:
     return (tuple(out_shape), output_placements, shardable_dims)
```
iiuc, we don't need to return the out shape anymore?

wanchaol commented on torch/distributed/_tensor/ops/view_ops.py:
```diff
@@ -631,121 +609,111 @@ def get_dim_size(cmd: DimSpec) -> Tuple[int, Optional[InputDim]]:
+                        _global_out_shape,
+                        shard_out,
+                        shardable_dims,
+                    ) = propagate_shape_and_sharding(
```
why we need to call this again?

wanchaol reviewed (commented):
first pass, this looks pretty good after an initial look, I do want to see if we can simplify `propagate_shape_and_sharding`, and why we need to call it twice

wanchaol commented:
looks like CI is failing with circular deps, please fix

tianyu-l committed:
Update on "[dtensor] refactor view ops to use OpStrategy"


As titled. Some ops require adjustment of output shape argument. In rule-based sharding prop, global output shape was inferred in the rule (in `view_ops.py`). In strategy-based sharding prop, it is now obtained from propagated out_tensor_meta (in `sharding_prop.py`).

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -161,6 +161,14 @@ def output_ndim(self):
         return self.strategies[0].output_spec.shape
 
+    @property
+    def ndim(self):
```
why we need to add these two properties, instead of just directly call `output_ndim` and `output_shape`?

tianyu-l replied:
This is to support the existing test in `text_view_ops.py`. As discussed offline, will send a follow up PR to deprecate `output_ndim` and `output_shape`.

wanchaol commented on test/distributed/_tensor/test_view_ops.py:
```diff
@@ -495,14 +494,14 @@ def test_complex_view_ops(self):
             Split(InputDim(1), (13, 2), 1),
         )
-        view_as_real_rule = ops[torch.view_as_real].dim_map(intermediate)
+        view_as_real_rule = dim_maps[torch.view_as_real](intermediate)
```
I think we'll need to improve some of the existing test cases in this file to work with CommDebugMode, to make sure the communication that are happening is expected. This can be done in a follow up PR

wanchaol reviewed (approved):
This looks great! Have some minor comments

tianyu-l committed:
Update on "[dtensor] refactor view ops to use OpStrategy"


As titled. Some ops require adjustment of output shape argument. In rule-based sharding prop, global output shape was inferred in the rule (in `view_ops.py`). In strategy-based sharding prop, it is now obtained from propagated out_tensor_meta (in `sharding_prop.py`).

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] refactor view ops to use OpStrategy"


As titled. Some ops require adjustment of output shape argument. In rule-based sharding prop, global output shape was inferred in the rule (in `view_ops.py`). In strategy-based sharding prop, it is now obtained from propagated out_tensor_meta (in `sharding_prop.py`).

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
